src/com/zone5cloud/retrofit/core/OkHttpClientInterceptor_Authorization.java

```diff
@@ -195,33 +202,44 @@ private void refreshIfRequired(Chain chain, String baseUrl) {
-																		 .add("username", username)
-																		 .add("refresh_token", token.getRefreshToken()).build();
-								String url = baseUrl + Users.NEW_ACCESS_TOKEN;
-								Request authRequest = new Request.Builder().url(url).post(body).build();
+							if (token.getRefreshToken() != null) {
+								// do a Cognito refresh
+								String username = token.extractUsername();
+//
+								if (username != null) {
+									// Add Api-Key and Api-Key-Secret to body if the request url matches zone5BaseUrl
+									if(originalRequestUrl.contains(zone5BaseUrl)) {
+										body = new FormBody.Builder()
+												.add("client_id", clientID)
+												.add("client_secret", clientSecret)
+												.add("grant_type", GrantType.REFRESH_TOKEN.toString())
+												.add("username", username)
+												.add("refresh_token", token.getRefreshToken()).build();
+									}else {
+										 body = new FormBody.Builder()
+												.add("grant_type", GrantType.REFRESH_TOKEN.toString())
+												.add("username", username)
+												.add("refresh_token", token.getRefreshToken()).build();
+									}
+									// assign the new token to original request Url
+									String url = originalRequestUrl + Users.NEW_ACCESS_TOKEN;
+									Request authRequest = new Request.Builder().url(url).post(body).build();
+									Response response = chain.proceed(authRequest);
+									saveNewToken(Users.NEW_ACCESS_TOKEN, response);
+									response.close();
+								}
+							} else {
+								// do a Gigya refresh
+								// note: I put a check for getRefreshToken() into the entry if block
+								// which will exclude this case from being hit. If we want to support Gigya token refresh
+								// remove the getRefreshToken() != null check from the entry if statement at the start of this method
+//								String url = baseUrl + Users.REFRESH_TOKEN;
+								String url = originalRequestUrl + Users.REFRESH_TOKEN;
+								Request authRequest = new Request.Builder().url(url).header(Z5HttpHeader.AUTHORIZATION.toString(), token.getBearer()).get().build();
 								Response response = chain.proceed(authRequest);
-								saveNewToken(Users.NEW_ACCESS_TOKEN, response);
+								saveNewToken(Users.REFRESH_TOKEN, response);
 								response.close();
 							}
-						} else {
-							// do a Gigya refresh
-							// note: I put a check for getRefreshToken() into the entry if block
-							// which will exclude this case from being hit. If we want to support Gigya token refresh
-							// remove the getRefreshToken() != null check from the entry if statement at the start of this method
-							String url = baseUrl + Users.REFRESH_TOKEN;
-							Request authRequest = new Request.Builder().url(url).header(Z5HttpHeader.AUTHORIZATION.toString(), token.getBearer()).get().build();
-							Response response = chain.proceed(authRequest);
-							saveNewToken(Users.REFRESH_TOKEN, response);
-							response.close();
-						}
-						
```

Comment:
This whole method refreshIfRequired(...) does not need changes - just the baseUrl being passed in needs to be the zone5 server, not the original request url.